### DISCUSSION TITLE
Logging elements dropped by ZStream.bufferSliding and ZStream.bufferDropping methods

### DISCUSSION THREAD

**Discussion by mpskowron** [zio/zio#5156] [Q&A]
Hi! How to log which elements were dropped by `ZStream.bufferSliding` or `ZStream.bufferDropping` methods?

**Comment by adamgfraser** [comment]
The `bufferSliding` and `bufferDropping` operators are designed for cases where you are okay potentially dropping arbitrary values.

If you need to take an action for every value that is dropped you can use a queue directly. For a dropping queue `offer` will return `false` if a value was not successfully offered so then you can log it and drop it. For a sliding queue you can use a `bounded` queue and if `offer` returns `false` call `take`, log the taken value and drop it, and then offer the new value again.

**Reply by mpskowron** [reply]
Thank you for an answer. In this case, I will probably just enrich each element with an index before `buffer*` method since I'm ok with dropping values - I just need to know if I start dropping too many.

**Reply by adamgfraser** [reply]
Yes, if you only need to know how many values were dropped using a monotonically increasing index is a good solution.
